coverage/lcovreport.py

```diff
@@ -0,0 +1,57 @@
+# Licensed under the Apache License: http://www.apache.org/licenses/LICENSE-2.0
+# For details: https://github.com/nedbat/coveragepy/blob/master/NOTICE.txt
+
+"""LCOV reporting for coverage.py"""
+
+import sys
+from hashlib import md5
+
+from coverage.report import get_analysis_to_report
+from coverage.exceptions import NoSource, NotPython
+
+
+class LcovReporter:
+    """A reporter for writing LCOV coverage reports"""
+
+    report_type = "LCOV report"
+
+    def __init__(self, coverage):
+        self.coverage = coverage
+        self.config = self.coverage.config
+
+    def report(self, morfs, outfile=None):
+        """Renders the full lcov report
+
+        'morfs' is a list of modules or filenames
+
+        outfile is the file object to write the file into
+        """
+
+        self.coverage.load()
+        self.coverage.get_data()
+        lcov_data = ""
+        outfile = outfile or sys.stdout
+        coverage_data = self.coverage.get_data()
+
+        for fr, analysis in get_analysis_to_report(self.coverage, morfs):
+            lcov_data += self.get_lcov(fr, analysis)
+        outfile.write(lcov_data)
+
+    def get_lcov(self, fr, analysis):
+        """Prooduces the lcov data for a single file"""
+        output = ""
+        token_lines = analysis.file_reporter.source().splitlines()
+        output += "TN:\n"
+        output += f"SF:{fr.relative_filename()}\n"
+
+        lines_hit = 0
+        for i, _ in enumerate(token_lines, 1):
+            if i in analysis.executed:
+                output += f"DA:{i},1\n"
+                lines_hit += 1
+            elif i in analysis.missing:
+                output += f"DA:{i},0\n"
+        output += f"LF:{len(analysis.statements)}\n"
+        output += f"LH:{lines_hit}\n"
```

Comment:
Counting iteratively while also iterating over a list may seem like it'd be faster, and it would be in something like C, but Python stores the length of a list so this is zero-cost and clearer:

```suggestion
        output += f"LH:{len(analysis.executed)}\n"
```

Reply:
Also note that I think `get_data()` returns this exact kind of
data. It'd be better to use that instead of re-calculating it here, in
case coveragepy changes its mind about how those statistics should be
calculated.